src/compiler/parser.ts

```diff
@@ -556,98 +558,111 @@ module ts {
         function setDecoratorContext(val: boolean) {
             setContextFlag(val, ParserContextFlags.Decorator);
         }
-
-        function doOutsideOfContext<T>(flags: ParserContextFlags, func: () => T): T {
-            let currentContextFlags = contextFlags & flags;
-            if (currentContextFlags) {
-                setContextFlag(false, currentContextFlags);
+        
+        function setAwaitContext(val: boolean) {
+            setContextFlag(val, ParserContextFlags.Await);
+        }
+        
+        function setAsyncParameterContext(val: boolean) {
+            setContextFlag(val, ParserContextFlags.AsyncParameter);
+        }
+        
+        function doOutsideOfContext<T>(context: ParserContextFlags, func: () => T): T {
+            let setContextFlags = context & contextFlags;

```

Comment:
'setContextFlags' is not a great name (esp as we have a function call setContextFlag).  Maybe 'newContextFlags'?